src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs

```diff
@@ -159,7 +200,49 @@ public void Execute(Script script)
                 throw new CakeException(message);
             }
 
-            roslynScript.RunAsync(_host).Wait();
+            if (_scriptCacheEnabled)
+            {
+                // Verify cache directory exists
+                if (!_fileSystem.GetDirectory(_scriptCachePath).Exists)
+                {
+                    _fileSystem.GetDirectory(_scriptCachePath).Create();
+                }
+                if (string.IsNullOrWhiteSpace(scriptHash))
+                {
+                    scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                }
+                var emitResult = compilation.Emit(cachedAssembly.FullPath);
+
+                if (emitResult.Success)
+                {
+                    using (var stream = _fileSystem.GetFile(hashFile).OpenWrite())
+                    using (var writer = new StreamWriter(stream, Encoding.UTF8))
+                    {
+                        writer.Write(scriptHash);
+                    }
+                    RunScriptAssembly(cachedAssembly.FullPath);
+                }
+            }
+            else
+            {
+                roslynScript.RunAsync(_host).Wait();
+            }
+        }
+
+        private void RunScriptAssembly(string assemblyPath)
+        {
+            var assembly = _loader.Load(assemblyPath, false);
+            var type = assembly.GetType("Submission#0");
+            var factoryMethod = type.GetMethod("<Factory>", new[] { typeof(object[]) });
+            try
+            {
+                var task = (Task<object>)factoryMethod.Invoke(null, new object[] { new object[] { _host, null } });
+                task.Wait();
```

Comment:
If we'd use `GetAwaiter().GetResult()` instead of `Wait()`, we could probably skip the `try / catch` which rethrows inner exception.

Reply:
Adjusted.